test/unit/throw.spec.js

```diff
@@ -25,6 +25,7 @@ describe('a test that throws', function() {
   });
 
   afterEach(function() {
+    process.removeAllListeners('uncaughtException');
```

Comment:
this starts to bite when running in watch mode

Reply:
I don't understand this comment.

Reply:
`Runner` will continually add `uncaughtException` listeners to `process` when running tests tests under mocha's "watch" mode.  we need to remove the listeners that we've created by calling `runner.run()`

Reply:
In `start()` the runner clears the `uncaughtException` handlers. So I don't see how these listeners could be accumulated. Even in watch mode (and even when aborting the execution before re-run), EVENT_RUN_END should be emitted and those handlers removed.
Our uncaught exception handling is an important part of our async testing. To remove this handler for testing seems strange to me. These tests are weird, but anyway ...

Reply:
> In run() the runner clears the uncaughtException handlers. So I don't see how these listeners could be accumulated. Even in watch mode (and even when aborting the execution before re-run), EVENT_RUN_END should be emitted and those handlers removed.

After `EVENT_RUN_END` is emitted, the `Runner#uncaught` listener is removed from the `Process.uncaughtException` event, _but_ the `Runner#uncaughtEnd` is then _added_ as a listener.  There is nothing that removes this listener _except_ another call to `Runner#run`.  

You'll see in these hooks that the listeners that the _Mocha process_ creates are restored after each test.  We have to remove them up-front because they conflict.  Any event listeners the _test subject_ adds (`Runner`) removed by the call to `removeAllListeners('uncaughtException')`, which--in the happy path--is just the `Runner#uncaughtEnd` listener.  I felt it was probably a better idea to remove any listeners the subject might add here, in case something weird happens.

Another way to approach this might be just stubbing out `process` at the module level via rewiremock.  I have no idea if that would work on a global, or if we'd need to `var process = require('process')`, or what.  Using sinon only to do this--stubbing `Process#on` and `Process#removeListener` would be even worse, because it's not sandboxed, and would break the mocha process which was running the test.

This brings up a good point though: we should not rely on `process#*` as globals.  It should be treated like `setTimeout` and its ilk.

I think it's best to probably revert the changes here and try to address these concerns in a separate PR, because it's more involved than I wanted to get with this one.

Reply:
nevermind about reverting--it's trivial to show the problem.  before my change, you see this (wherein in called `touch test/unit/throw.spec.js` to trigger the run):

```
$ bin/mocha --watch test/unit/throw.spec.js



  a test that throws
    non-extensible
      ✓ should not pass if throwing sync and test is sync
      ✓ should not pass if throwing sync and test is async
      ✓ should not pass if throwing async and test is async
    undefined
      ✓ should not pass if throwing sync and test is sync
      ✓ should not pass if throwing sync and test is async
      ✓ should not pass if throwing async and test is async
    null
      ✓ should not pass if throwing sync and test is sync
      ✓ should not pass if throwing sync and test is async
      ✓ should not pass if throwing async and test is async


  9 passing (27ms)



  a test that throws
    non-extensible
      ✓ should not pass if throwing sync and test is sync
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
      ✓ should not pass if throwing sync and test is async
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
      ✓ should not pass if throwing async and test is async
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
    undefined
      ✓ should not pass if throwing sync and test is sync
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
      ✓ should not pass if throwing sync and test is async
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
      ✓ should not pass if throwing async and test is async
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
    null
      ✓ should not pass if throwing sync and test is sync
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
      ✓ should not pass if throwing sync and test is async
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
      ✓ should not pass if throwing async and test is async
(node:96463) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit


  9 passing (22ms)
```

Reply:
_each_ test of `Runner#run` adds a listener (`Runner#uncaughEnd`).  when we run this the second time, we quickly go over the max listeners limit of 10.

Reply:
this is not noticeable in watch mode otherwise, because the `Runner` instance is reused, and it cleans up after itself upon a second call.